src/nvim/os/win_defs.h

```diff
@@ -1,24 +1,181 @@
 #ifndef NVIM_OS_WIN_DEFS_H
 #define NVIM_OS_WIN_DEFS_H
 
+#include "auto/config.h"
+
+#ifndef LC_MESSAGES
+// FIXME: this is a hack - also used in ex_cmds2.c
+// to build in windows without libintl
+# define LC_MESSAGES LC_COLLATE
+#endif
+
+// winsock2.h must be before windows.h - or so says Mingw
+#include <winsock2.h>
 #include <windows.h>
+// We have our own RGB macro in macros.h
+#undef RGB
+
+#include <io.h>
+#include <uv.h>
+
+#include <stdio.h>
+#include <time.h>
+#include <sys/stat.h>
+
+// FIXME: this is only used by if_cscope.c
+#ifndef S_ISLNK
+# define S_ISLNK(mode) 0
+#endif
+
+// For MSVC
+#ifdef _MSC_VER
+# ifndef restrict
+#  define restrict __restrict
+# endif
+# ifndef inline
+#  define inline __inline
+# endif
+# ifndef __func__
+#  define __func__ __FUNCTION__
+# endif
+# ifndef STDOUT_FILENO
+#  define STDOUT_FILENO _fileno(stdout)
+# endif
+# ifndef STDERR_FILENO
+#  define STDERR_FILENO _fileno(stderr)
+# endif
+# ifndef W_OK
+// There is no W_OK for MSVC but it is 02
+# define W_OK 2
```

Comment:
W_OK (as well as R_OK) and other flags are defined in libuv's `include/uv-win.h`. I believe we should include `include/uv-win.h` in our `win_defs.h` since it defines the unix-equivalent flags required by libuv's functions.